allennlp/modules/token_embedder.py

```diff
@@ -0,0 +1,38 @@
+from typing import Dict
+
+import torch
+
+from allennlp.common import Params
+from allennlp.data import Vocabulary
+
+class TokenEmbedder(torch.nn.Module):
+    """
+    A ``TokenEmbedder`` is a ``Module`` that takes as input the :class:`~allennlp.data.DataArray`
+    produced by a :class:`~allennlp.data.fields.TextField` and returns as output an embedded
+    representation of the tokens in that field.
+
+    The ``DataArrays`` produced by ``TextFields`` are `dictionaries` with named representations,
+    like "words" and "characters".  When you create a ``TextField``, you pass in a dictionary of
+    :class:`~allennlp.data.TokenIndexer` objects, telling the field how exactly the tokens in the
+    field should be represented.  This class changes the type signature of ``Module.forward``,
+    restricting ``TokenEmbedders`` to take inputs corresponding to a single ``TextField``.  We also
+    add a method to the basic ``Module`` API: :func:`get_output_dim()`.  You might need this if you
+    want to construct a ``Linear`` layer using the output of this embedder, for instance.
+    """
+    def forward(self,  # pylint: disable=arguments-differ
+                text_field_input: Dict[str, torch.Tensor]) -> torch.Tensor:
+        raise NotImplementedError
+
+    def get_output_dim(self) -> int:
```

Comment:
So you're proposing something like:

```
"text_field_embedder": {
  "output_dim": 10,
  "token_embedders": {
    "words": {
      ...
      "embedding_dim": 4
    },
    "characters": {
      "output_dim": 6,
      ...
    }
  }
}
```

And then we check in the code that the computed output dim matches the specified output dim?  If we don't have that check, we have a crash later in a seemingly unrelated part of the code, wherever the output tensor gets used.

Seems simpler to just have it like this:

```
"text_field_embedder": {
  "type": "basic"  // this line optional, because "basic" is the default
  "words": {
    ...
    "embedding_dim": 4
  },
  "characters": {
    "output_dim": 6,
    ...
  }
}
```

And then we compute the output dim for you, so you don't have to repeat yourself.  If one parameter is computable from another one, you generally shouldn't be specifying it twice.

The alternative I was thinking about was dropping the `get_output_shape()` method entirely, and requiring `input_dim` to be specified everywhere.  This is already true for encoders built `from_params`, but it isn't true for `Linear` layers or other things that are built in the model constructor.  The choice here is between `get_output_shape()` as a method on embedders and encoders, and taking various `input_dims` as model parameters for `Linear` or other layers (like attentions) that we want to construct given the encoders and embedders we're using.  I think `get_output_shape()` is cleaner.